File: tests/qt_types_standalone/rust/src/qrectf.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QRectF;

#[cxx::bridge]
mod qrectf_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/include/qt_types.h");

        type QRectF = cxx_qt_lib::QRectF;
    }

    extern "Rust" {
        fn construct_qrectf() -> QRectF;
        fn read_qrectf(p: &QRectF) -> bool;
        fn clone_qrectf(p: &QRectF) -> QRectF;
    }
}

fn construct_qrectf() -> QRectF {
    QRectF::new(1.23, 4.56, 2.46, 9.12)
}

fn read_qrectf(p: &QRectF) -> bool {
    ((p.x() - 1.23).abs() < f64::EPSILON)
        && ((p.y() - 4.56).abs() < f64::EPSILON)
        && ((p.width() - 2.46).abs() < f64::EPSILON)
        && ((p.height() - 9.12).abs() < f64::EPSILON)
}

fn clone_qrectf(p: &QRectF) -> QRectF {
    p.clone()
}

File: crates/cxx-qt-lib/src/core/qrectf.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qrectf.h"

#include "../assertion_utils.h"

// QRectF has 4 double members
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qrect.h?h=v5.15.6-lts-lgpl#n621
//
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qrect.h?h=v6.2.4#n623
constexpr static ::std::array<::std::size_t, 4> arr{ sizeof(double),
                                                     sizeof(double),
                                                     sizeof(double),
                                                     sizeof(double) };
assert_alignment_and_size(QRectF, alignof(double), arr, arr.size());

static_assert(::std::is_trivially_copyable<QRectF>::value,
              "QRectF must be trivially copyable");
